# Preprocessing

In [177]:
import pandas as pd
import numpy as np

In [178]:
from functions import *

### Import the datasets

In [179]:
# Open the raw datasets
# Heavy Metal
hm_df = pd.read_excel('data/AllHMFiles.xlsx', engine='openpyxl').fillna(0.0)

# Pesticides
pesticides_df = pd.read_excel('data/AllPesticidesFiles.xlsx', engine='openpyxl').fillna(0.0)
classe_pesticides_df = pd.read_excel('data/pesticides.xlsx', engine='openpyxl')

# Surfaces
surfaces_df = pd.read_excel('data/absSurfs.xlsx', engine='openpyxl')

# Distances
distances_df = pd.read_excel('data/distsOneSheet.xlsx', engine='openpyxl')

In [180]:
hm_df.head(5)

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [181]:
pesticides_df.head(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Nicosulfuron  \
0           0.0           0.0          0.0        0.0  ...           0.0   
1           0.0           0.0          0.0        0.0  ...           0.0   
2           0.0           0.0          0.0        0.0  ...           0.0   
3           0.0           0.0          0.0        0.0  ...           0.0   
4           0.0           0.0          0.0        0.0  ...           0.0   

   Piperonylbutoxide  Prosulfuron  Simazine  Terbuthylazine  Dieldrin  \
0                0.0          0.0       0.0             0.0       0.0   
1                0.0          0.0       0.0             0.0       0.0   
2                0.0          0.0       0.0             0.0       0.0   
3                0.0          0.0       0.0             0.0       0.0   
4                0.0          0.0       0.0             0.0       0.0   

   Fenamidone  Trifluralin  Cycloate  Pentachloroanisol  
0         0.0          0.0       0.0                0.0  
1         0.0          0.0       0.0                0.0  
2         0.0          0.0       0.0                0.0  
3         0.0          0.0       0.0                0.0  
4         0.0          0.0       0.0                0.0  

[5 rows x 117 columns]

In [182]:
classe_pesticides_df.head(5)

importName           name    statusEN   LMR  \
0      Aclonifen      aclonifen  authorized  0.05   
1         Aldrin         aldrin  prohibited  0.01   
2   Ametoctradin   ametoctradin  authorized  0.05   
3  Anthraquinone  anthraquinone  prohibited  0.02   
4   Azoxystrobin   azoxystrobin  authorized  0.05   

                          familyEN          typeEN  
0                   diphenyl ether       herbicide  
1                   organochlorine     insecticide  
2               triazolopyrimidine       fungicide  
3  polycyclic aromatic hydrocarbon  bird repellant  
4                      strobilurin       fungicide

In [183]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [184]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

### Heavy Metal

In [185]:
hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [186]:
MRL_HM_dict = {
    'Lead' : [1.5],
    'Cadmium' : [1],
    'Arsenic' : [0.3],
    'Mercury' : [1],
    'Copper' : [30]
}

MRL_HM = pd.DataFrame.from_dict(MRL_HM_dict)
MRL_HM

Lead  Cadmium  Arsenic  Mercury  Copper
0   1.5        1      0.3        1      30

In [187]:
hm_df['mrl_lead'] = np.where(hm_df['Lead'] > 1.5, 1, 0)
hm_df['mrl_cadmium'] = np.where(hm_df['Cadmium'] > 1, 1, 0)
hm_df['mrl_arsenic'] = np.where(hm_df['Arsenic'] > 0.3, 1, 0)
hm_df['mrl_mercury'] = np.where(hm_df['Mercury'] > 1, 1, 0)
hm_df['mrl_copper'] = np.where(hm_df['Copper'] > 30, 1, 0)

hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
0     0.032    40     0.0         0            0            0            0   
1     0.029    33     0.0         0            0            0            0   
2     0.038    37     0.0         0            0            0            0   
3     0.031    37     0.0         0            0            0            0   
4     0.052    35     0.0         0            0            0            0   

   mrl_copper  
0           0  
1           0  
2           0  
3           0  
4           0

In [188]:
#hm_df = hm_df.drop(['Chromium'], axis=1)
hm_df = hm_df.drop(['Chrome'], axis=1)
hm_df = hm_df.drop(['Zinc'], axis=1)

hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  mrl_copper  
0     0.032         0            0            0            0           0  
1     0.029         0            0            0            0           0  
2     0.038         0            0            0            0           0  
3     0.031         0            0            0            0           0  
4     0.052         0            0            0            0           0

### Pesticides

Fungicide :

In [189]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].head(5)

importName                       name    statusEN   LMR  \
2         Ametoctradin               ametoctradin  authorized  0.05   
4         Azoxystrobin               azoxystrobin  authorized  0.05   
5            Benalaxyl                  benalaxyl  prohibited  0.05   
7  Benthiavalicarb.iso  benthiavalicarb-isopropyl  authorized  0.05   
8     Benzovindiflupyr           benzovindiflupyr  authorized  0.05   

             familyEN     typeEN  
2  triazolopyrimidine  fungicide  
4         strobilurin  fungicide  
5        phenylamides  fungicide  
7           carbamate  fungicide  
8               amide  fungicide

In [190]:
fungicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].importName.to_list()
fungicides.append('Diphenylamine')
fungicides.append('Tebuconazole')
fungicides

['Ametoctradin',
 'Azoxystrobin',
 'Benalaxyl',
 'Benthiavalicarb.iso',
 'Benzovindiflupyr',
 'Boscalid',
 'Bupirimate',
 'Captan',
 'Carbendazim',
 'Chlorothalonil',
 'Cyazofamid',
 'Cyflufenamid',
 'Cymoxanil',
 'Cyprodinil',
 'Difenoconazole',
 'Dimethomorph',
 'Dimoxystrobin',
 'Dodine',
 'Epoxiconazole',
 'Ethirimol',
 'Famoxadone',
 'Fenamidone',
 'Fenbuconazole',
 'Fenhexamid',
 'Fenpropidin',
 'Fenpropimorph',
 'Fluazinam',
 'Fludioxonil',
 'Fluopicolide',
 'Fluopyram',
 'Flutolanil',
 'Fluxapyroxad',
 'Hexaconazole',
 'Iprovalicarb',
 'Isofetamid',
 'Mandipropamid',
 'Metalaxyl',
 'Metconazole',
 'Metrafenone',
 'Penconazole',
 'Pencycuron',
 'Pentachloroanisol',
 'Prochloraz',
 'Propamocarb',
 'Propiconazole',
 'Prothioconazole_d',
 'Pyraclostrobin',
 'Pyrimethanil',
 'Quinoxyfen',
 'Spiroxamine',
 'Tetraconazole',
 'Thiabendazole',
 'Thiophanate.methyl',
 'Trifloxystrobin',
 'X2.phenylphenol',
 'Zoxamide',
 'Diphenylamine',
 'Tebuconazole']

Insecticide :

In [191]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].head(5)

importName                 name          statusEN   LMR  \
1                Aldrin               aldrin        prohibited  0.01   
13  Chlorantraniliprole  chlorantraniliprole        authorized  0.05   
17  Chlorpyrifos.methyl  chlorpyrifos-methyl        prohibited  0.05   
19            Coumaphos            coumaphos        prohibited  0.10   
20         Cyanofenphos         cyanofenphos  not yet assessed   NaN   

               familyEN       typeEN  
1        organochlorine  insecticide  
13  anthranilic diamide  insecticide  
17      organophosphate  insecticide  
19        thiophosphate  insecticide  
20      organophosphate  insecticide

In [192]:
insecticides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].importName.to_list()
insecticides.append('Diphenylamine')
insecticides.append('Methiocarb_total')
insecticides

['Aldrin',
 'Chlorantraniliprole',
 'Chlorpyrifos.methyl',
 'Coumaphos',
 'Cyanofenphos',
 'Cyflumetofen',
 'Cypermethrin',
 'Deet',
 'Dieldrin',
 'Etofenprox',
 'Fenoxycarb',
 'Fenpropathrin',
 'Flonicamid',
 'Flucythrinate',
 'Flupyradifurone',
 'Imidacloprid',
 'Indoxacarb',
 'Methoxyfenozide',
 'Phosmet',
 'Piperonylbutoxide',
 'Pirimicarb',
 'Spinosad',
 'Spirotetramat',
 'Tau.fluvalinate',
 'Tebufenozide',
 'Tetradifon',
 'Thiacloprid',
 'Diphenylamine',
 'Methiocarb_total']

Herbicide : 

In [193]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].head(5)

importName           name    statusEN   LMR          familyEN  \
0       Aclonifen      aclonifen  authorized  0.05    diphenyl ether   
6     Benfluralin    benfluralin  authorized  0.05    dinitroaniline   
15  Chlorotoluron  chlorotoluron  authorized  0.05        phenylurea   
16   Chlorpropham   chlorpropham  prohibited  0.05         carbamate   
18      Clethodim      clethodim  authorized  0.05  cyclohexanedione   

       typeEN  
0   herbicide  
6   herbicide  
15  herbicide  
16  herbicide  
18  herbicide

In [194]:
herbicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].importName.to_list()
herbicides

['Aclonifen',
 'Benfluralin',
 'Chlorotoluron',
 'Chlorpropham',
 'Clethodim',
 'Cycloate',
 'Diflufenican',
 'Ethofumesate',
 'Fluazifop',
 'Flufenacet',
 'Metamitron',
 'Methabenzthiazuron',
 'Metobromuron',
 'Metolachlor',
 'Nicosulfuron',
 'Pendimethalin',
 'Phenmedipham',
 'Propanil',
 'Propyzamide',
 'Prosulfocarb',
 'Prosulfuron',
 'Simazine',
 'Terbuthylazine',
 'Tri.allate',
 'Trifluralin']

Others :

In [195]:
classe_pesticides_df[(classe_pesticides_df['typeEN'] != 'fungicide') & 
                     (classe_pesticides_df['typeEN'] != 'insecticide') & 
                     (classe_pesticides_df['typeEN'] != 'herbicide')]

importName           name    statusEN   LMR  \
3           Anthraquinone  anthraquinone  prohibited  0.02   
34          Diphenylamine  diphenylamine  prohibited  0.05   
69       Methiocarb_total     methiocarb  prohibited  0.05   
86             Propargite     propargite  prohibited  0.05   
100          Tebuconazole   tebuconazole  authorized  0.05   
112  X6.benzylaminopurine  benzyladenine  authorized   NaN   

                            familyEN  \
3    polycyclic aromatic hydrocarbon   
34                             amine   
69                         carbamate   
86                               NaN   
100                         triazole   
112                       cytokinins   

                                           typeEN  
3                                  bird repellant  
34    growth regulator, fungicide and insecticide  
69   insecticide, molluscicide and bird repellent  
86                                      acaricide  
100                fungicide and growth regulator  
112                              growth regulator

##### Choice to treat the others

Adding to the corresponding categories, stay in "other" ? What about the growth regulator ?

#### Add some columns

In [196]:
pesticides_df['qtt_fungicides'] = pesticides_df.apply(qtt_pesticides, args = fungicides, axis = 1)
pesticides_df['qtt_insecticides'] = pesticides_df.apply(qtt_pesticides, args = insecticides, axis = 1)
pesticides_df['qtt_herbicides'] = pesticides_df.apply(qtt_pesticides, args = herbicides, axis = 1)

In [197]:
pesticides_df['nb_fungicides'] = pesticides_df.apply(nb_pesticides, args = fungicides, axis = 1)
pesticides_df['nb_insecticides'] = pesticides_df.apply(nb_pesticides, args = insecticides, axis = 1)
pesticides_df['nb_herbicides'] = pesticides_df.apply(nb_pesticides, args = herbicides, axis = 1)

In [198]:
pesticides_df['lmr_fungicides'] = pesticides_df.apply(lmr_pesticides, args = fungicides, axis = 1)
pesticides_df['lmr_insecticides'] = pesticides_df.apply(lmr_pesticides, args = insecticides, axis = 1)
pesticides_df['lmr_herbicides'] = pesticides_df.apply(lmr_pesticides, args = herbicides, axis = 1)

In [199]:
pesticides_df.head(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   
5            B0039  2017/P1     0.0     0.0000          0.00           0.0   
6            B0041  2017/P1     0.0     0.0000          0.00           0.0   
7            B0043  2017/P1     0.0     0.0000          0.00           0.0   
8            B0045  2017/P1     0.0     0.0000          0.00           0.0   
9            B0059  2017/P1     0.0     0.0000          0.00           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Pentachloroanisol  \
0          0.00           0.0         0.00        0.0  ...                0.0   
1          0.00           0.0         0.00        0.0  ...                0.0   
2          0.00           0.0         0.00        0.0  ...                0.0   
3          0.00           0.0         0.00        0.0  ...                0.0   
4          0.00           0.0         0.00        0.0  ...                0.0   
5          0.01           0.0         0.00        0.0  ...                0.0   
6          0.00           0.0         0.00        0.0  ...                0.0   
7          0.00           0.0         0.01        0.0  ...                0.0   
8          0.00           0.0         0.00        0.0  ...                0.0   
9          0.00           0.0         0.00        0.0  ...                0.0   

   qtt_fungicides  qtt_insecticides  qtt_herbicides  nb_fungicides  \
0          0.0000            0.0000            0.00              0   
1          0.0188            0.0000            0.00              1   
2          0.0711            0.0000            0.00              1   
3          0.0294            0.0000            0.00              1   
4          0.0000            0.0000            0.01              0   
5          0.0000            0.0208            0.00              0   
6          0.0000            0.0000            0.00              0   
7          0.0100            0.0000            0.00              1   
8          0.0000            0.0000            0.00              0   
9          0.0145            0.0000            0.00              1   

   nb_insecticides  nb_herbicides  lmr_fungicides  lmr_insecticides  \
0                0              0               0                 0   
1                0              0               0                 0   
2                0              0               1                 0   
3                0              0               0                 0   
4                0              1               0                 0   
5                2              0               0                 0   
6                0              0               0                 0   
7                0              0               0                 0   
8                0              0               0                 0   
9                0              0               0                 0   

   lmr_herbicides  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  
7               0  
8               0  
9               0  

[10 rows x 126 columns]

In [200]:
pesticides_df['year'] = pesticides_df['PERIOD'].apply(lambda x: x[:4])
pesticides_df['trimester'] = pesticides_df['PERIOD'].apply(lambda x: x[-1:])
pesticides_df.sample(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
422            F0031  2020/P2     0.0       0.00           0.0           0.0   
210            B0105  2019/P1     0.0       0.00           0.0           0.0   
184            B0061  2018/P4     0.0       0.00           0.0           0.0   
232            F0041  2019/P1     0.0       0.00           0.0           0.0   
58             B0096  2017/P3     0.0       0.00           0.0           0.0   
491            B0093  2020/P4     0.0       0.00           0.0           0.0   
459            F0005  2020/P3     0.0       0.00           0.0           0.0   
444            B0043  2020/P3     0.0       0.00           0.0           0.0   
352            B0093  2020/P1     0.0       0.01           0.0           0.0   
427            F0057  2020/P2     0.0       0.00           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
422           0.0           0.0          0.0     0.0000  ...   
210           0.0           0.0          0.0     0.0105  ...   
184           0.0           0.0          0.0     0.0000  ...   
232           0.0           0.0          0.0     0.0414  ...   
58            0.0           0.0          0.0     0.0000  ...   
491           0.0           0.0          0.0     0.0064  ...   
459           0.0           0.0          0.0     0.0000  ...   
444           0.0           0.0          0.0     0.0000  ...   
352           0.0           0.0          0.0     0.0150  ...   
427           0.0           0.0          0.0     0.0000  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
422             0.093          0.0000              1                1   
210             0.000          0.0503              3                0   
184             0.000          0.0000              0                0   
232             0.000          0.0000              2                0   
58              0.000          0.0000              0                0   
491             0.000          0.0000              2                0   
459             0.000          0.0000              0                0   
444             0.000          0.0000              2                0   
352             0.000          0.0000              3                0   
427             0.000          0.0000              0                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
422              0               1                 1               0  2020   
210              2               0                 0               0  2019   
184              0               0                 0               0  2018   
232              0               0                 0               0  2019   
58               0               0                 0               0  2017   
491              0               0                 0               0  2020   
459              0               0                 0               0  2020   
444              0               0                 0               0  2020   
352              0               0                 0               0  2020   
427              0               0                 0               0  2020   

     trimester  
422          2  
210          1  
184          4  
232          1  
58           3  
491          4  
459          3  
444          3  
352          1  
427          2  

[10 rows x 128 columns]

##### Alternate (smaller) dataset for pesticides

In [201]:
pesti_names = pesticides_df.columns.to_list()[1:-11]

In [202]:
small_pesticides_df = pesticides_df.drop(pesti_names, axis=1)
small_pesticides_df.head(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
0            B0017          0.0000               0.0            0.00   
1            B0019          0.0188               0.0            0.00   
2            B0021          0.0711               0.0            0.00   
3            B0023          0.0294               0.0            0.00   
4            B0037          0.0000               0.0            0.01   

   nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
0              0                0              0               0   
1              1                0              0               0   
2              1                0              0               1   
3              1                0              0               0   
4              0                0              1               0   

   lmr_insecticides  lmr_herbicides  year trimester  
0                 0               0  2017         1  
1                 0               0  2017         1  
2                 0               0  2017         1  
3                 0               0  2017         1  
4                 0               0  2017         1

##### Alternate (group by) dataset for pesticides

In [203]:
group_pesti_df = pesticides_df.groupby('REF....SUBSTANCE').agg(max)

In [204]:
group_pesti_df = group_pesti_df.loc[:,'lmr_fungicides' : 'lmr_herbicides']
group_pesti_df

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
B0012                          1                 0               0
B0017                          1                 0               0
B0019                          0                 0               0
B0021                          1                 0               0
B0023                          0                 0               0
B0032                          1                 0               0
B0037                          1                 1               0
B0039                          1                 0               0
B0041                          1                 0               0
B0043                          1                 0               1
B0045                          1                 0               0
B0059                          1                 0               0
B0061                          0                 0               0
B0067                          1                 0               0
B0069                          1                 1               0
B0077                          1                 1               0
B0093                          1                 0               0
B0096                          0                 0               0
B0103                          1                 0               0
B0105                          1                 1               1
B0116                          0                 0               0
B0128                          0                 0               0
B0140                          0                 0               0
B0142                          1                 0               0
B0152                          1                 0               0
B0158                          1                 0               0
B0163                          1                 0               0
F0001                          1                 1               0
F0003                          1                 0               0
F0005                          0                 0               0
F0007                          0                 0               0
F0009                          0                 0               0
F0011                          0                 1               0
F0013                          0                 0               0
F0015                          1                 0               0
F0017                          0                 0               0
F0019                          1                 0               0
F0020                          1                 0               0
F0021                          1                 0               0
F0023                          0                 0               1
F0025                          0                 0               0
F0029                          1                 0               0
F0031                          1                 1               0
F0033                          0                 0               0
F0035                          0                 0               1
F0039                          1                 0               0
F0041                          0                 0               0
F0043                          0                 0               0
F0055                          1                 0               0
F0057                          1                 0               0
F0061                          1                 0               0
F0063                          1                 0               0
F0065                          0                 0               0
F0069                          0                 0               1
F0071                          1                 0               0
F0073                          1                 1               0
FE0001                         0                 0               0

### Sites

##### Getting the area for each zone

In [205]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

In [206]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [207]:
# surfaces_df['country'] = surfaces_df['Site'].apply(lambda x: 'FR' if x.startswith('F') else 'BE')
# surfaces_df['country'].value_counts()

In [208]:
sites_df = distances_df

In [209]:
sites_df['surface'] = sites_df.apply(get_surface, axis = 1)

In [210]:
sites_df.sample(10)

Site     polyID  classCLC         dist       surface
175  F0067  FR-226510       312   646.001843  6.412933e+05
337  F0029   FR-24970       112  1274.996080  2.943795e+05
23   F0027  FR-201661       311   303.870953  3.212519e+06
146  F0043   FR-29150       121   689.907493  7.608398e+05
239  B0043   BE-11811       231   227.418700  1.000232e+06
572  F0009  FR-220153       312   577.036456  2.851793e+05
465  F0039  FR-200725       311   457.842778  5.577001e+05
435  B0105    BE-2397       112   804.818038  6.810172e+05
559  F0009  FR-122258       242   532.430146  1.979240e+06
545  F0015  FR-187751       311   102.771173  5.107448e+06

In [211]:
sites_df['country'] = sites_df['polyID'].apply(lambda x: 'FR' if x.startswith('FR') else 'BE')
sites_df.sample(10)

Site     polyID  classCLC         dist       surface country
560  F0009   FR-81414       231     0.000000  2.907895e+06      FR
469  F0039   FR-28091       121   851.360822  4.008746e+05      FR
454  B0103   BE-18176       242   769.404874  2.385882e+06      BE
208  B0045   BE-24952       523  1217.224329  2.535093e+05      BE
597  B0069    BE-6852       211  1352.353418  3.496322e+06      BE
453  B0103   BE-14469       311   396.084173  5.394297e+05      BE
594  B0069    BE-1392       231   381.929259  2.521990e+05      BE
38   F0023  FR-211615       311     0.000000  2.938245e+06      FR
298  B0093   BE-21188       311   186.426297  3.876686e+05      BE
543  F0013  FR-220123       312   198.211960  1.271223e+06      FR

In [212]:
sites_df['country'].value_counts()

FR    335
BE    297
Name: country, dtype: int64

In [213]:
sites_df = sites_df.drop(['polyID'], axis=1)
sites_df.sample(5)

Site  classCLC         dist       surface country
1    F0069       242   289.926324  2.278667e+06      FR
226  B0037       242  1392.939449  1.335673e+06      BE
124  F0019       211  1239.116211  1.588808e+06      FR
438  B0105       211     0.000000  5.559084e+06      BE
409  B0021       231  1237.725789  2.695177e+05      BE

##### Get one dataset with : beehives and area/distance for each zone

In [214]:
surfaces_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [215]:
classCLC = surfaces_df.columns.to_list()[1:]

In [216]:
by_sites_df = surfaces_df.copy(deep=True)
by_sites_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [217]:
for CLC in classCLC:
    by_sites_df[CLC] = by_sites_df.apply(distance_area, args = (CLC,), axis = 1)

In [218]:
by_sites_df

Site  111           112           121         122  123  124  131  132  \
0   F0069  0.0  1.442780e+03  3.587087e+05    0.000000  0.0  0.0  0.0  0.0   
1   F0025  0.0  0.000000e+00  0.000000e+00  700.801763  0.0  0.0  0.0  0.0   
2   F0027  0.0  2.604161e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
4   F0023  0.0  0.000000e+00  7.907522e+02    0.000000  0.0  0.0  0.0  0.0   
..    ...  ...           ...           ...         ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
59  B0069  0.0  7.328884e+03  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
60  B0152  0.0  4.575279e+06  5.227880e+02  421.402885  0.0  0.0  0.0  0.0   
61  F0003  0.0  3.859675e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
62  B0158  0.0  1.574508e+02  0.000000e+00  392.706036  0.0  0.0  0.0  0.0   

    133  ...  411  412  421  422  423  511         512  521  522  523  
0   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
1   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
2   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
4   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  463.344615  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...         ...  ...  ...  ...  
58  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
59  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
60  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
61  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
62  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[63 rows x 45 columns]

check if correct

In [219]:
distance = distances_df[(distances_df['Site'] == 'F0023') & 
                        (distances_df['classCLC'] == 512)]['dist'].values
distance

array([614.50674791])

In [220]:
area = surfaces_df[surfaces_df['Site'] == 'F0023']['512'].values[0]
area

284728.392578125

In [221]:
area/distance[0]

463.3446150839585

### Merging the datasets into two

3 datasets to 2 : sites + (heavy metal || pesticides)

##### Heavy Metal

##### Pesticides

full

In [222]:
pesticides_df.sample(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
82             B0039  2018/P1     0.0        0.0           0.0           0.0   
380            F0063  2020/P1     0.0        0.0           0.0           0.0   
448            B0069  2020/P3     0.0        0.0           0.0           0.0   
310            B0017  2019/P4     0.0        0.0           0.0           0.0   
235            B0017  2019/P2     0.0        0.0           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
82            0.0           0.0       0.0000     0.0000  ...   
380           0.0           0.0       0.0000     0.0271  ...   
448           0.0           0.0       0.0046     0.7077  ...   
310           0.0           0.0       0.0000     0.0000  ...   
235           0.0           0.0       0.0000     0.0000  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
82              0.000          0.0000              0                0   
380             0.011          0.0305              7                2   
448             0.000          0.0000              9                0   
310             0.000          0.0000              0                0   
235             0.000          0.0000              0                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
82               0               0                 0               0  2018   
380              4               1                 0               0  2020   
448              0               1                 0               0  2020   
310              0               0                 0               0  2019   
235              0               0                 0               0  2019   

     trimester  
82           1  
380          1  
448          3  
310          4  
235          2  

[5 rows x 128 columns]

In [223]:
small_pesticides_df.sample(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
373            F0031          0.0561               0.0           0.000   
6              B0041          0.0000               0.0           0.000   
502            F0011          0.0000               0.0           0.000   
292            B0103          0.0637               0.0           0.000   
368            F0019          2.1429               0.0           0.005   

     nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
373              4                0              0               0   
6                0                0              0               0   
502              0                0              0               0   
292              3                0              0               0   
368              6                0              1               1   

     lmr_insecticides  lmr_herbicides  year trimester  
373                 0               0  2020         1  
6                   0               0  2017         1  
502                 0               0  2020         4  
292                 0               0  2019         3  
368                 0               0  2020         1

In [224]:
group_pesti_df.sample(5)

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
F0023                          0                 0               1
B0152                          1                 0               0
F0041                          0                 0               0
F0039                          1                 0               0
B0142                          1                 0               0

##### sites

In [225]:
sites_df.sample(5)

Site  classCLC         dist       surface country
600  B0069       242   952.401356  1.933157e+06      BE
170  F0067       313   794.331161  2.955444e+05      FR
94   B0142       242  1078.693783  1.750115e+05      BE
255  F0035       313  1125.133485  2.258232e+05      FR
465  F0039       311   457.842778  5.577001e+05      FR

In [226]:
by_sites_df.sample(5)

Site  111           112           121         122  123  124        131  \
5   F0071  0.0      0.000000  2.918492e+02    0.000000  0.0  0.0  57.312693   
24  B0037  0.0  12746.310883  0.000000e+00    0.000000  0.0  0.0   0.000000   
3   B0128  0.0   2412.961618  4.053755e+06  523.436355  0.0  0.0   0.000000   
55  F0005  0.0    524.046563  0.000000e+00    0.000000  0.0  0.0   0.000000   
50  F0063  0.0    339.909994  0.000000e+00    0.000000  0.0  0.0   0.000000   

    132  133  ...  411  412  421  422  423  511         512  521  522  523  
5   0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  631.183641  0.0  0.0  0.0  
24  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
55  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
50  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[5 rows x 45 columns]

#### Merge

In [227]:
group_pesti_df.index.name

'REF....SUBSTANCE'

In [228]:
group_pesti_df = group_pesti_df.reset_index()

In [229]:
group_pesti_df = group_pesti_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [230]:
group_pesti_df.head(3)

Site  lmr_fungicides  lmr_insecticides  lmr_herbicides
0  B0012               1                 0               0
1  B0017               1                 0               0
2  B0019               0                 0               0

In [231]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  132  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   133  ...  411  412  421  422  423  511  512  521  522  523  
0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[3 rows x 45 columns]

In [232]:
merged_dataset = pd.merge(group_pesti_df, by_sites_df, on=["Site"])

In [233]:
merged_dataset

Site  lmr_fungicides  lmr_insecticides  lmr_herbicides           111  \
0    B0012               1                 0               0    209.227976   
1    B0017               1                 0               0      0.000000   
2    B0019               0                 0               0      0.000000   
3    B0021               1                 0               0      0.000000   
4    B0023               0                 0               0      0.000000   
5    B0032               1                 0               0      0.000000   
6    B0037               1                 1               0      0.000000   
7    B0039               1                 0               0   3098.057557   
8    B0041               1                 0               0    391.077976   
9    B0043               1                 0               1      0.000000   
10   B0045               1                 0               0      0.000000   
11   B0059               1                 0               0      0.000000   
12   B0061               0                 0               0      0.000000   
13   B0067               1                 0               0      0.000000   
14   B0069               1                 1               0      0.000000   
15   B0077               1                 1               0      0.000000   
16   B0093               1                 0               0      0.000000   
17   B0096               0                 0               0      0.000000   
18   B0103               1                 0               0      0.000000   
19   B0105               1                 1               1      0.000000   
20   B0116               0                 0               0      0.000000   
21   B0128               0                 0               0      0.000000   
22   B0140               0                 0               0   1416.095618   
23   B0142               1                 0               0      0.000000   
24   B0152               1                 0               0      0.000000   
25   B0158               1                 0               0      0.000000   
26   B0163               1                 0               0  15107.362039   
27   F0001               1                 1               0      0.000000   
28   F0003               1                 0               0      0.000000   
29   F0005               0                 0               0      0.000000   
30   F0007               0                 0               0      0.000000   
31   F0009               0                 0               0      0.000000   
32   F0011               0                 1               0      0.000000   
33   F0013               0                 0               0      0.000000   
34   F0015               1                 0               0      0.000000   
35   F0017               0                 0               0      0.000000   
36   F0019               1                 0               0      0.000000   
37   F0020               1                 0               0      0.000000   
38   F0021               1                 0               0      0.000000   
39   F0023               0                 0               1      0.000000   
40   F0025               0                 0               0      0.000000   
41   F0029               1                 0               0      0.000000   
42   F0031               1                 1               0      0.000000   
43   F0033               0                 0               0      0.000000   
44   F0035               0                 0               1      0.000000   
45   F0039               1                 0               0      0.000000   
46   F0041               0                 0               0      0.000000   
47   F0043               0                 0               0      0.000000   
48   F0055               1                 0               0      0.000000   
49   F0057               1                 0               0      0.000000   
50   F0061               1  

In [234]:
merged_dataset.to_csv('merged_dataset.csv')

### Changing the NA

# Questions to ask

- When NaN => information irrelevant or not analysed?
- Difference France/Belgium?
- Is the fact that an insecticide is unauthorized relevant? (=> not the same law in France/Belgium)
- Or do we have to show that even banned pesticides are still found?
- How do we treat other categorical pesticides?
- How to increase the number of datas?

### Bo-bo-bonus

- FR  :  36 beehives for 335 zones
- BE  :  27 beehives for 297 zones

Here the CLC zone are left separate, could group some to test